Météo-France platform for the weather component

victorcerutti opened the pull request:
This is reopening of https://github.com/home-assistant/home-assistant/pull/18322 after I removed the weather model change I made.
Discussions about enhancing the weather component (model and frontend) have been posted to :
- https://github.com/home-assistant/architecture/issues/102
- https://github.com/home-assistant/architecture/issues/75

## Description:
After the recently merged Météo-France sensor providing weather datas from Météo-France, here is the `weather` component
The configuration is similar to the weather sensor and the weather platform also provides a 4 days forecast.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7480

## Example entry for `configuration.yaml` (if applicable):
```yaml
weather:
  - platform: meteo_france
    city: '76000'
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

victorcerutti committed:
new weather component for meteofrance

victorcerutti committed:
linting

victorcerutti committed:
upgrade meteofrance package version

victorcerutti committed:
Update .coveragerc

victorcerutti committed:
Remove updates to the weather component architecture

fabaff commented on homeassistant/components/weather/meteo_france.py:
```diff
@@ -0,0 +1,145 @@
+from homeassistant.helpers import config_validation as cv
+# Reuse data and API logic from the sensor implementation
+from homeassistant.components.sensor.meteo_france import \
+    MeteoFranceUpdater, CONF_POSTAL_CODE, CONF_ATTRIBUTION
```
Unfortunately we allowed this in the past...I'm sorry, today you need to create a `meteo_france` component and make the `weather` and the `sensor` platform depend on the component which contain the logic to retrieve the data.

victorcerutti replied:
Thanks for the review
I will update the PR in a few days with an appropriate solution

fabaff reviewed (dismissed):
At first glance looks the rest pretty solid.

MartinHjelmare commented on homeassistant/components/weather/meteo_france.py:
```diff
@@ -0,0 +1,145 @@
+})
+
+CONDITION_CLASSES = {
+    'clear-night': ['Nuit Claire'],
```
Probably make the keys lowercase snakecase.

victorcerutti committed:
Rewrite Météo-France as a component

victorcerutti commented:
Please consider my last commit as a work in progress.
I wanted to check if I'm doing this correctly
This is a **breaking change** for the sensor as the configuration is now done this way:
```yaml
meteo_france:
  - postal_code: 29000
    monitored_conditions:
        - rain_chance
        - freeze_chance
        - thunder_chance
        - snow_chance
        - weather
        - next_rain
        - wind_speed
        - temperature
        - uv
    weather_card: false
  - postal_code: 76000
    monitored_conditions:
      - temperature
  - postal_code: 51100
```

For each component, we display by default a weather card (unless config says otherwise) and a sensor for each monitored conditions (optional)

All the logic of client and update is managed by the `meteo_france` component.
Sensors and weather card only refers to a single client data fetcher for each city.

If this is correct, I will continue in this direction and update the documentation to reflect this changes

victorcerutti committed:
Update .coveragerc

victorcerutti committed:
Update requirements_all.txt

victorcerutti commented:
@fabaff Can you please review the way the config is set for this component ?
If this is ok I will continue and update the documentation and the PR to introduce this new component.

victorcerutti commented:
Hello,
I'd love to have feedback on this PR regarding the configuration here https://github.com/home-assistant/home-assistant/pull/18404#issuecomment-438714097

Many thanks for considering my request !

balloob commented:
We should not have a weather card config option, it's up to the user with Lovelace to decide if they want to show a weather card for an entity or not.

The config seems fine.

victorcerutti commented:
Thank you for the feedback
It makes total sense as Lovelace is now default to discard the `weather_card` option
I will correct this and update this PR soon

victorcerutti committed:
remove Weather Card option

victorcerutti committed:
Update conf name

Changing conf name to something more universal for worldwide weather forecast (postal code was only relevent for France)

victorcerutti committed:
Update meteofrance pypi package

victorcerutti committed:
fix line too long

victorcerutti committed:
Update requirements_all.txt

victorcerutti commented:
I think this is all good
I also updated the documentation according to the new configuration of the component

victorcerutti committed:
prevent from calling an API endpoint if not in monitored conditions

victorcerutti committed:
fix stale url and remove blank line

victorcerutti committed:
Insure that all cities are unique

victorcerutti committed:
rename CONF_ATTRIBUTION

victorcerutti committed:
Updating data from component setup

victorcerutti committed:
fix missing extra lines

victorcerutti commented on homeassistant/components/meteo_france.py:
```diff
@@ -0,0 +1,130 @@
+Support for Meteo France weather forecast.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.meteo_france/
```
Thanks, it's fixed now

victorcerutti commented on homeassistant/components/meteo_france.py:
```diff
@@ -0,0 +1,130 @@
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [vol.Schema({
+        vol.Required(CONF_CITY): cv.string,
```
Thank you for pointing out a working solution.
I integrated a similar check to insure that all cities are unique

victorcerutti commented on homeassistant/components/meteo_france.py:
```diff
@@ -0,0 +1,134 @@
+    def __init__(self, client):
+        """Initialize the data object."""
+        self._client = client
+        self.update()
```
fixed

MartinHjelmare reviewed (approved):
Looks great!

MartinHjelmare commented:
Can be merged when build passes.

anthosz commented:
Dear @fabaff @mxworm ,

Do you know in which version this component will be available?

Best regards,

MartinHjelmare commented:
0.89
